File: infrastructure/spring-boot-webflux-rest-api-adapter/src/main/java/codes/fdk/blueprint/api/infrastructure/rest/webflux/CommandMapper.java
package codes.fdk.blueprint.api.infrastructure.rest.webflux;

import codes.fdk.blueprint.api.domain.command.CreateCategoryCommand;
import codes.fdk.blueprint.api.domain.command.UpdateCategoryCommand;
import codes.fdk.blueprint.api.domain.model.Category;
import codes.fdk.blueprint.api.domain.model.CategoryId;
import org.mapstruct.Mapper;

import static codes.fdk.blueprint.api.domain.command.CreateCategoryCommand.createChildCategoryCommand;
import static codes.fdk.blueprint.api.domain.command.CreateCategoryCommand.createRootCategoryCommand;

//TODO figure out how to let mapstruct do the work
@Mapper
interface CommandMapper {

    default CreateCategoryCommand toCreateCommand(PostCategoryRequest request) {
        return createRootCategoryCommand(
                request.name(),
                request.slug(),
                request.isVisible()
        );
    }

    default CreateCategoryCommand toCreateCommand(CategoryId parentId, PostCategoryRequest request) {
        return createChildCategoryCommand(
                request.name(),
                request.slug(),
                parentId,
                request.isVisible()
        );
    }

    default UpdateCategoryCommand toUpdateCommand(CategoryId id, PatchCategoryRequest request) {
        return new UpdateCategoryCommand(id, request.isVisible());
    }

    default GetCategoryResponse toResponse(Category category) {
        return new GetCategoryResponse(
                category.id(),
                category.name(),
                category.slug(),
                category.parentId(),
                category.isVisible()
        );
    }

}

File: infrastructure/vertx-web-openapi-adapter/src/main/kotlin/codes/fdk/blueprint/api/infrastructure/web/openapi/CommandMapper.kt
package codes.fdk.blueprint.api.infrastructure.web.openapi

import codes.fdk.blueprint.api.domain.command.CreateCategoryCommand
import codes.fdk.blueprint.api.domain.command.UpdateCategoryCommand
import codes.fdk.blueprint.api.domain.model.CategoryId

internal object CommandMapper {

    fun toCreateCommand(request: PostCategoryRequest): CreateCategoryCommand {
        return CreateCategoryCommand.createRootCategoryCommand(
            request.name,
            request.slug,
            request.visible
        )
    }

    fun toCreateCommand(parentId: CategoryId, request: PostCategoryRequest): CreateCategoryCommand {
        return CreateCategoryCommand.createChildCategoryCommand(
            request.name,
            request.slug,
            parentId,
            request.visible
        )
    }

    fun toUpdateCommand(id: CategoryId, request: PatchCategoryRequest): UpdateCategoryCommand {
        return UpdateCategoryCommand(id, request.visible)
    }

}